language-examples/remedial-js/9.0-functional-programming/README.md

```diff
@@ -0,0 +1,238 @@
+# Functional Programming
+
+Functional Programming simply means composing programs with series' of small, reusable, **pure functions**.
+
+## Pure Functions 
+
+A pure function is a function that, given the same input, will always return the same output, and has no side-effects.
+
+```javascript
+function double (a) {
+  return a + a
+}
+
+// boom
+double(5) // 10
+```
+
+### Side Effects
+For a function to be pure, it must have absolutely no side-effects. It cannot access, mutate, or depend upon shared state in any capacity. This includes mutating params, and accessing any mutable state that has not been explicitly passed to the function.
+
+
+```javascript
+let globalVar = 4
+
+// impure, depends on shared state, unpredictable output
+function impureAdd (a) {
+  return a + globalVar
+}
+
+// impure, mutates shared state
+function impureDivide (a) {
+  globalVar /= a
+}
+
+// impure, mutates param
+function impureParamSub (obj, b) {
+  obj.val -= b
+}
+
+// pure, returns new object rather than mutating old one
+function pureParamSub (obj, b) {
+  return Object.assign({}, obj, {
+    val: obj.val -b
+  })
+}
+
+// pure, no dependencies, predictable output
+function pureAdd (a, b) {
+  return a + b
+}
+```
+
+Functions like `Date.now()` and `Math.random()` (and any function that depends on them) are considered impure because their outputs are not predictable.
+
+```javascript
+// impure unpredictable output
+function impureCreateObj (objName) {
+  return {
+    name: objName,
+    id: Math.random() * 100000
+  }
+}
+
+// passing in the unique id allows this function to remain pure
```

Comment:
Yeah, I wasn't sure how to clarify that without saying the same thing again. I added one more example below it (`console.log` one), hopefully it's clear now.